draft-ietf-quic-qpack.md

```diff
@@ -127,23 +127,180 @@ considered blocked by the decoder and can not be processed until all entries in
+instruction space:
+
+ - Table updates are carried by HEADERS frames on the control stream, as defined
+   by {{QUIC-HTTP}}.  Frames on this stream modify the dynamic table state
+   without generating output to any particular request.
+ - Acknowledgement of header frame processing is carried by HEADER_ACK frames on
+   the control stream, running from decoder to encoder.
+ - Finally, the contents of HEADERS and PUSH_PROMISE frames on request streams
+   reference the QPACK table state.
+
+This section describes the instructions which are possible on each stream type.
+
+In order to ensure table consistency and simplify update management, all table
+updates occur on the control stream rather than on request streams. Request
+streams contain only header blocks, which do not modify the state of the table.
+
+## Primitives
+
+The prefixed integer from Section 5.1 of [RFC7541] is used heavily throughout
+this document.  The string literal, defined by Section 5.2 of [RFC7541], is used
+with the following modification.
+
+HPACK defines string literals to begin on a byte boundary.  They begin with a
+single flag (indicating whether the string is Huffman-coded), followed by the
+Length encoded as a 7-bit prefix integer, and finally Length octets of data.
+
+QPACK permits strings to begin other than on a byte boundary.  An "N-bit prefix
+string literal" begins with the same Huffman flag, followed by the length
+encoded as an (N-1)-bit prefix integer.  The remainder of the string literal is
+unmodified.
+
+A string literal without a prefix length noted is an 8-bit prefix string literal
+and follows the definitions in [RFC7541] without modification.
+
+## HEADERS Frames on the Control Stream
+
+Table updates can add a table entry, possibly using existing entries to avoid
+transmitting redundant information.  The name can be transmitted as a reference
+to an existing entry in the static or the dynamic table or as a string literal.
+For entries which already exist in the dynamic table, the full entry can also be
+used by reference, creating a duplicate entry.
+
+### Insert With Name Reference
+
+An addition to the header table where the header field name matches the header
+field name of an entry stored in the static table or the dynamic table starts
+with the '1' one-bit pattern.  The `S` bit indicates whether the reference is to
+the static (S=1) or dynamic (S=0) table. The header field name is represented
+using the index of that entry, which is represented as an integer with a 6-bit
+prefix (see Section 5.1 of [RFC7541]). Table indices are always non-zero; a zero
+index MUST be treated as a decoding error.
+
+The header name reference is followed by the header field value represented as a
+string literal (see Section 5.2 of [RFC7541]).
+
+~~~~~~~~~~ drawing
+     0   1   2   3   4   5   6   7
+   +---+---+---+---+---+---+---+---+
+   | 1 | S |    Name Index (6+)    |
+   +---+---+-----------------------+
```

Comment:
May I ask why we need to define the index as a tuple of `S` bit and an 6+ index, rather than just a 7+ index?

I am not sure why doing `(octet & 0x40) != 0` is better than `index > 62`, and IIUC splitting the index into two by `S` reduces the number of dynamic table indexes that can be crammed into a single octet from 65 to 62.

Other than that, I think that this is a fantastic PR.

Reply:
IIRC, the idea is to be able to grow the static table -- or even to switch it for an alternative.  For example, a static table ID could be sent as transport parameters.

Reply:
Basically.  #904 tracks improvements to the static table, which could be as simple as defining a new one (probably based on some httparchive data discussed on that issue) or as complex as a negotiation scheme to permit selecting one on each connection.  By moving to the bit, we insulate the dynamic table performance from the possibility that the future static table will be larger.

It also permits us to simplify processing by removing sentinel values (index zero is special).  For the moment, it also allows us to drop the static bit in the Duplicate instruction, but #1314 improves on that further, so I wouldn't count that a large benefit unless we don't take that PR.

Reply:
@dtikhonov @MikeBishop Thank you for the answers.

While I do not think that hard-coding the split the space into two using a single bit is ideal (ideally I think all static indexes should be given greater values (which are probably encoded longer) than the dynamic ones considering the fact that a static table is a pre-learnt dictionary), I can see that such encoding is hard to handle especially in case of QPACK.

I now agree that using a single bit is a nice compromise between simplicity, efficiency, extensibility.